src/decisionengine/framework/dataspace/datablock.py

```diff
@@ -245,6 +246,16 @@ def __str__(self):
     def __contains__(self, key):
         return key in self.keys()
 
+    @property
+    def _keys(self):
+        self.logger.debug("datablock waiting for internal read lock in '_keys'")
+        with self.__internal_data_read_lock:
+            return tuple(self.dataspace.get_datablock(self.sequence_id, self.generation_id).keys())
+
+    @_keys.setter
+    def _keys(self, value):  # pragma: no cover
+        raise ValueError("You may not redefine the known keys for a datablock")
+
```

Comment:
You changed it already to be a tuple retrieved from the DB.
Could this be removed (leaving only .keys() )?
The only file using ._keys is test_datablock.py which could be modified to use .keys()
If kept it could become a cache to avoid multiple queries to the DB. I did not check how many times this is needed, so not sure if it is worth it.
Another question (in the opposite direction) is that the DB was more an archival component, not used for run time. This removes 2 elements (keeping only one) that should be consistent and may become inconsistent but seems to introduce a run time dependency from the DB.